hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/PartialDisconnectionHandler.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.cluster.impl;
+
+import com.hazelcast.cluster.impl.MemberImpl;
+import com.hazelcast.internal.util.graph.BronKerboschCliqueFinder;
+import com.hazelcast.internal.util.graph.Graph;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static java.util.Collections.emptySet;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+class PartialDisconnectionHandler {
+    private final long detectionIntervalMs;
+    private final long algorithmTimeoutMs;
+    private Map<MemberImpl, Set<MemberImpl>> disconnections = new HashMap<>();
+    private long lastUpdated;
+
+    PartialDisconnectionHandler(long detectionIntervalMs, long algorithmTimeoutMs) {
+        this.detectionIntervalMs = detectionIntervalMs;
+        this.algorithmTimeoutMs = algorithmTimeoutMs;
+    }
+
+    boolean update(MemberImpl member, long timestamp, Collection<MemberImpl> disconnectedMembers) {
+        if (timestamp < lastUpdated) {
+            return false;
+        }
+
+        Set<MemberImpl> currentDisconnectedMembers =  disconnections.get(member);
+        if (currentDisconnectedMembers == null) {
+            if (disconnectedMembers.isEmpty()) {
+                return false;
+            }
+
+            currentDisconnectedMembers = new HashSet<>();
+            disconnections.put(member, currentDisconnectedMembers);
+        }
+
+        boolean updated = false;
+
+        for (MemberImpl disconnectedMember : disconnectedMembers) {
+            if (currentDisconnectedMembers.add(disconnectedMember)
```

Comment:
Type of `currentDisconnectedMembers ` is a `Set`, as i understand it is because same `MemberImpl` can be added multiple times?